topology.go

```diff
@@ -272,7 +276,87 @@ func (gs *generationSourcePre4_4) getGeneration(genTime time.Time, consistency g
 }
 
 func (gs *generationSourcePre4_4) getGenerationTimes(consistency gocql.Consistency) ([]time.Time, error) {
-	iter := gs.session.Query("SELECT time FROM " + generationsTableNamePre4_4).
+	iter := gs.session.Query("SELECT time FROM " + systemDistributedKeyspace + "." + generationsTableNamePre4_4).
+		Consistency(consistency).
+		Iter()
+	var (
+		times    []time.Time
+		currTime time.Time
+	)
+	for iter.Scan(&currTime) {
+		times = append(times, currTime)
+	}
+	if err := iter.Close(); err != nil {
+		return nil, err
+	}
+	return times, nil
+}
+
+// Follows the migration procedure from Scylla's documentation
+// https://docs.scylladb.com/using-scylla/cdc/cdc-querying-streams/
+func (gs *generationSourcePre4_4) maybeUpgrade(consistency gocql.Consistency) (generationSource, error) {
+	meta, err := gs.session.KeyspaceMetadata(systemDistributedKeyspace)
+	if err != nil {
+		return gs, err
+	}
+
+	// Check if table is present
+	_, ok := meta.Tables[streamsTableSince4_4]
+	if !ok {
+		// Don't upgrade, the new table is not there yet
+		return gs, nil
+	}
+
+	// Was the migration completed?
+	data := make(map[string]interface{})
+	err = gs.session.Query("SELECT streams_timestamp FROM system.cdc_local WHERE key = 'rewritten'").
+		Consistency(consistency).
+		MapScan(data)
+
+	if err == gocql.ErrNotFound {
+		// The "rewritten" row is not present yet, this means that the generations
+		// weren't rewritten yet
+		// Try again later
+		return gs, nil
+	}
+
+	if err != nil {
+		// Some other error
+		return gs, err
+	}
+
+	newGs := &generationSourceSince4_4{
+		session: gs.session,
+		logger:  gs.logger,
+	}
+
+	return newGs.maybeUpgrade(consistency)
+}
+
+type generationSourceSince4_4 struct {
+	session *gocql.Session
+	logger  Logger
+}
+
+func (gs *generationSourceSince4_4) getGeneration(genTime time.Time, consistency gocql.Consistency) ([]StreamID, error) {
+	var streams []StreamID
+	iter := gs.session.Query("SELECT streams FROM "+systemDistributedKeyspace+"."+streamsTableSince4_4+" WHERE time = ?", genTime).
+		Consistency(consistency).
+		Iter()
+
+	var vnodeStreams []StreamID
```

Comment:
BTW. with this new table you don't need to group stream IDs manually at a later stage (I assume you do somewhere based on the stream ID format) - this is the grouping, already given. Perhaps the getGeneration API could return stream IDs already grouped, so this new implementation could be better.
But meh, it doesn't matter that much, it's a rare op.

Reply:
Yes, streams are grouped per vnode using the `(*Reader).splitStreams` function from `reader.go`.

I'd rather leave the `getGeneration` API as it is - the stream grouping algorithm is simple, It's rarely invoked, and `getGeneration` API will stay simpler that way.